File: src/trail-renderer.js
class TrailRenderer {
  constructor(
    value71,
    other,
    extra,
    extra2,
    extra3,
    extra4,
    value72 = 16777215,
    value73 = 1,
  ) {
    ((this["_color"] = value72),
      (this["_opacity"] = value73),
      (this["_fadeDelta"] = 1 / extra),
      (this["_minSegSq"] = extra2 * extra2),
      (this["_maxSeg"] = extra4),
      (this["_maxPoints"] = 5 * Math["floor"](60 * extra + 2)),
      (this["_stroke"] = extra3),
      (this["_pts"] = []),
      (this["_posR"] = {
        x: 0,
        y: 0,
      }),
      (this["_posInit"] = !1),
      (this["_active"] = !1),
      (this["_gfx"] = value71["add"]["graphics"]()),
      this["_gfx"]["setBlendMode"](Phaser["BlendModes"]["ADD"]));
  }
  ["addToContainer"](value71, other) {
    (value71["add"](this["_gfx"]), this["_gfx"]["setDepth"](other));
  }
  ["setPosition"](value71, other) {
    ((this["_posR"]["x"] = value71),
      (this["_posR"]["y"] = other),
      (this["_posInit"] = !0));
  }
  ["start"]() {
    this["_active"] = !0;
  }
  ["stop"]() {
    this["_active"] = !1;
  }
  ["reset"]() {
    ((this["_pts"] = []), (this["_posInit"] = !1), this["_gfx"]["clear"]());
  }
  ["update"](value71) {
    if (!this["_posInit"]) return void this["_gfx"]["clear"]();
    const value72 = value71 * this["_fadeDelta"];
    let num32 = 0;
    for (let i = 0; i < this["_pts"]["length"]; i++)
      ((this["_pts"][i]["state"] -= value72),
        this["_pts"][i]["state"] > 0 &&
          (num32 !== i && (this["_pts"][num32] = this["_pts"][i]), num32++));
    if (
      ((this["_pts"]["length"] = num32),
      this["_active"] && this["_pts"]["length"] < this["_maxPoints"])
    ) {
      const length2 = this["_pts"]["length"];
      let value73 = !0;
      if (length2 > 0) {
        const point = this["_pts"][length2 - 1],
          value74 = this["_posR"]["x"] - point["x"],
          value75 = this["_posR"]["y"] - point["y"],
          value76 = value74 * value74 + value75 * value75;
        if (this["_maxSeg"] > 0 && Math["sqrt"](value76) > this["_maxSeg"])
          this["_pts"]["length"] = 0;
        else if (value76 < this["_minSegSq"]) value73 = !1;
        else if (length2 > 1) {
          const point2 = this["_pts"][length2 - 2],
            value77 = this["_posR"]["x"] - point2["x"],
            value78 = this["_posR"]["y"] - point2["y"];
          value77 * value77 + value78 * value78 < 2 * this["_minSegSq"] &&
            (value73 = !1);
        }
      }
      value73 &&
        this["_pts"]["push"]({
          x: this["_posR"]["x"],
          y: this["_posR"]["y"],
          state: 1,
        });
    }
    this["_gfx"]["clear"]();
    const length = this["_pts"]["length"];
    if (!(length < 2))
      for (let i = 0; i < length - 1; i++) {
        const i2 = this["_pts"][i],
          point = this["_pts"][i + 1],
          value73 = 0.5 * (i2["state"] + point["state"]) * this["_opacity"];
        (this["_gfx"]["lineStyle"](this["_stroke"], this["_color"], value73),
          this["_gfx"]["lineBetween"](
            i2["x"],
            i2["y"],
            point["x"],
            point["y"],
          ));
      }
  }
}

File: src/trail-renderer.cpp
#include "trail-renderer.h"

TrailRenderer::TrailRenderer(float lifetime, float minSeg, float stroke,
                             float maxSeg, unsigned int color, float opacity)
: _color(color),
_opacity(opacity),
_fadeDelta(lifetime > 0.0f ? 1.0f / lifetime : 1.0f),
_minSegSq(minSeg * minSeg),
_maxSeg(maxSeg),
_maxPoints(5 * (int)std::floor(60.0f * lifetime + 2.0f)),
_stroke(stroke),
_posInit(false),
_active(false)
{
}

TrailRenderer::~TrailRenderer() {}

void TrailRenderer::setPosition(float x, float y) {
    _posX = x;
    _posY = y;
    _posInit = true;
}

void TrailRenderer::start() {
    _active = true;
}

void TrailRenderer::stop() {
    _active = false;
}

void TrailRenderer::reset() {
    _pts.clear();
    _posInit = false;
}

void TrailRenderer::update(float dt) {
    if (!_posInit) {
        _pts.clear();
        return;
    }

    float delta = dt * _fadeDelta;
    size_t validCount = 0;

    for (size_t i = 0; i < _pts.size(); ++i) {
        _pts[i].state -= delta;
        if (_pts[i].state > 0.0f) {
            if (validCount != i) {
                _pts[validCount] = _pts[i];
            }
            validCount++;
        }
    }
    _pts.resize(validCount);

    if (_active && (int)_pts.size() < _maxPoints) {
        size_t len = _pts.size();
        bool shouldAdd = true;

        if (len > 0) {
            const auto& last = _pts[len - 1];
            float dx = _posX - last.x;
            float dy = _posY - last.y;
            float distSq = dx * dx + dy * dy;

            if (_maxSeg > 0.0f && std::sqrt(distSq) > _maxSeg) {
                _pts.clear();
            } else if (distSq < _minSegSq) {
                shouldAdd = false;
            } else if (len > 1) {
                const auto& secondLast = _pts[len - 2];
                float dx2 = _posX - secondLast.x;
                float dy2 = _posY - secondLast.y;
                if (dx2 * dx2 + dy2 * dy2 < 2.0f * _minSegSq) {
                    shouldAdd = false;
                }
            }
        }

        if (shouldAdd) {
            _pts.push_back({_posX, _posY, 1.0f});
        }
    }
}

void TrailRenderer::render(float cameraX, float cameraY) {
    if (_pts.size() < 2) return;

    float r = ((_color >> 16) & 0xFF) / 255.0f;
    float g = ((_color >> 8)  & 0xFF) / 255.0f;
    float b =  (_color        & 0xFF) / 255.0f;

    glPushMatrix();
    glTranslatef(-cameraX, cameraY, 0.0f);

    glDisable(GL_TEXTURE_2D);
    applyBlendMode(BLEND_ADD);
    glLineWidth(_stroke);

    glBegin(GL_LINES);
    for (size_t i = 0; i + 1 < _pts.size(); ++i) {
        const auto& p1 = _pts[i];
        const auto& p2 = _pts[i + 1];
        float alpha = 0.5f * (p1.state + p2.state) * _opacity;

        glColor4f(r, g, b, alpha);
        glVertex2f(p1.x, p1.y);
        glVertex2f(p2.x, p2.y);
    }
    glEnd();

    glLineWidth(1.0f);
    applyBlendMode(BLEND_NORMAL);
    glEnable(GL_TEXTURE_2D);

    glPopMatrix();
}
